BasicHistorySessionsKotlin/app/src/main/java/com/google/android/gms/fit/samples/basichistorysessionskotlin/MainActivity.kt

```diff
@@ -0,0 +1,513 @@
+import com.google.android.gms.fitness.data.Field
+import com.google.android.gms.fitness.data.Session
+import com.google.android.gms.fitness.request.DataDeleteRequest
+import com.google.android.gms.fitness.request.SessionInsertRequest
+import com.google.android.gms.fitness.request.SessionReadRequest
+import com.google.android.gms.fitness.result.SessionReadResponse
+import com.google.android.gms.tasks.Task
+import com.google.android.material.snackbar.Snackbar
+import java.util.Calendar
+import java.util.Date
+import java.util.TimeZone
+import java.util.concurrent.TimeUnit
+
+const val TAG = "BasicSessions"
+const val SAMPLE_SESSION_NAME = "Afternoon run"
+
+/**
+ * This sample demonstrates how to use the Sessions API of the Google Fit platform to insert
+ * sessions into the History API, query against existing data, and remove sessions. It also
+ * demonstrates how to authenticate a user with Google Play Services and how to properly
+ * represent data in a Session, as well as how to use ActivitySegments.
+ */
+class MainActivity : AppCompatActivity() {
+    private val fitnessOptions: FitnessOptions by lazy {
+        FitnessOptions.builder()
+                .addDataType(DataType.TYPE_ACTIVITY_SEGMENT, FitnessOptions.ACCESS_WRITE)
+                .addDataType(DataType.TYPE_SPEED, FitnessOptions.ACCESS_WRITE)
+                .build()
+    }
+
+    // Two maps are used in conjunction with the OAuth flow. This is in order to identify what
+    // function to execute once the flow has successfully completed.
+    private val signInActionMap = mapOf<() -> Any, Int>(
+            ::insertAndVerifySession to 1,
+            ::deleteSession to 2
+    )
+    private val signInResponseMap = signInActionMap.entries.associate { (k, v) -> v to k }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        // This method sets up our custom logger, which will print all log messages to the device
+        // screen, as well as to adb logcat.
+        initializeLogging()
+
+        // When permissions are revoked the app is restarted so onCreate is sufficient to check for
+        // permissions core to the Activity's functionality.
+
+        checkPermissionsAndRun(::insertAndVerifySession)
+    }
+
+    private fun checkPermissionsAndRun(signInAction: () -> Any) {
+        if (hasRuntimePermissions()) {
+            fitSignIn(signInAction)
+        } else {
+            requestRuntimePermissions(signInAction)
+        }
+    }
+
+    /**
+     * Checks that the user is signed in, and if so, executes the specified function. If the user is
+     * not signed in, initiates the sign in flow, specifying the post-sign in function to execute.
+     *
+     * @param signInAction The function to execute if already signed in, or once signed in.
+     */
+    private fun fitSignIn(signInAction: () -> Any) {
+        if (!hasOAuthPermissions()) {
```

Comment:
As called out above, make this a positive and it's easier to read...

if (hasOAuthPermissions()) {
...